File: java-server/gate/src/main/java/com/mir2/gate/CharacterDescription.java
package com.mir2.gate;

import com.mir2.protocol.SixBitCodec;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;

/** Legacy packed TCharDesc: Feature(Integer) + Status(Integer), little-endian then 6-bit encoded. */
public record CharacterDescription(int feature, int status) {
  public static final int BYTES = 8;

  public String encode() {
    byte[] bytes = ByteBuffer.allocate(BYTES).order(ByteOrder.LITTLE_ENDIAN)
        .putInt(feature).putInt(status).array();
    return new String(SixBitCodec.encode(bytes), StandardCharsets.ISO_8859_1);
  }

  public static CharacterDescription decode(String encoded) {
    byte[] bytes = SixBitCodec.decodeString(encoded);
    if (bytes.length != BYTES) throw new IllegalArgumentException("TCharDesc must decode to 8 bytes");
    ByteBuffer buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
    return new CharacterDescription(buffer.getInt(), buffer.getInt());
  }
}

File: mir2-front/bridge/src/protocol/CharacterDescription.ts
import { SixBitCodec } from './SixBitCodec.js';

/**
 * Legacy packed TCharDesc: Feature(Integer) + Status(Integer), little-endian then 6-bit encoded (8 bytes -> encoded string).
 */
export class CharacterDescription {
  public static readonly BYTES = 8;

  constructor(
    public readonly feature: number,
    public readonly status: number
  ) {}

  public encode(): string {
    const buf = Buffer.alloc(CharacterDescription.BYTES);
    buf.writeInt32LE(this.feature, 0);
    buf.writeInt32LE(this.status, 4);
    return SixBitCodec.encodeString(buf);
  }

  public static decode(encoded: string): CharacterDescription {
    const bytes = SixBitCodec.decodeString(encoded);
    if (bytes.length < CharacterDescription.BYTES) {
      throw new Error(`TCharDesc must decode to at least 8 bytes, got ${bytes.length}`);
    }
    return new CharacterDescription(
      bytes.readInt32LE(0),
      bytes.readInt32LE(4)
    );
  }
}
